### PR TITLE
Modify some details on the build

### LINKED ISSUE #83: Add icons to the repository and bring Windows compatibility
At the moment the icons are download from an NPM archive. It isn't bad but it requires that the machine has curl and tar to download and extract the file. Since it might not be really relevant to download the file from its origin, we could simply add the file to the repository.

Code to change: `yewprint/build.rs`

- [ ] Add the icons to the repository
- [x] Add Windows test to the CI

### PR DESCRIPTION
**Pull request by yozhgoor**
closes #83

### CODE REVIEW

**Comment by yozhgoor** [comment]
Build for windows is good i thing
cc @cecton

**Comment by cecton** [comment]
> Build for windows is good i thing
> cc @cecton

If the build for Windows was good it would fail because it doesn't compile on Windows at the moment lol

File: yewprint/build.rs [hunk 1]
```diff
@@ -4,21 +4,11 @@ use std::collections::HashSet;
...
-        .arg(format!(
-            "curl -sSLo - {} | tar xOzf - {}",
-            ICON_URL, ICON_PATH
-        ))
-        .output()
-        .map(|x| String::from_utf8(x.stdout).expect("source file is not UTF-8"))
-        .expect("could not download icons");
+    let icon_svg_paths = fs::read("iconSvgPaths.js")
+        .map(|x| String::from_utf8(x).expect("source file is not UTF-8"))
+        .expect("could not read file");
```
Comment (cecton) [inline yewprint/build.rs:11]: Even shorter: https://doc.rust-lang.org/std/fs/fn.read_to_string.html

File: yewprint/build.rs [hunk 2]
```diff
@@ -4,21 +4,9 @@ use std::collections::HashSet;
 use std::env;
 use std::fs;
 use std::path::Path;
-use std::process::Command;
-
-const ICON_URL: &str = "https://registry.npmjs.org/@blueprintjs/icons/-/icons-3.19.0.tgz";
-const ICON_PATH: &str = "package/lib/esnext/generated/iconSvgPaths.js";
```
Comment (cecton) [inline yewprint/build.rs:10]: we should keep that in comment so people know how to update

Comment (yozhgoor) [reply yewprint/build.rs:10]: Not a bad idea, i fix that right now

File: yewprint/build.rs [hunk 3]
```diff
@@ -4,21 +4,12 @@ use std::collections::HashSet;
 use std::env;
 use std::fs;
 use std::path::Path;
-use std::process::Command;
 
-const ICON_URL: &str = "https://registry.npmjs.org/@blueprintjs/icons/-/icons-3.19.0.tgz";
-const ICON_PATH: &str = "package/lib/esnext/generated/iconSvgPaths.js";
+// Icon_url = "https://registry.npmjs.org/@blueprintjs/icons/-/icons-3.19.0.tgz";
+// Icon_path = "package/lib/esnext/generated/iconSvgPaths.js";
```
Comment (cecton) [inline yewprint/build.rs:9]: I doubt someone who's new to the code will understand what to do with this. Can you make a sentence in plain English explaining this?

File: yewprint/build.rs [hunk 4]
```diff
@@ -4,21 +4,12 @@ use std::collections::HashSet;
...
-    let icon_svg_paths = Command::new("sh")
-        .arg("-c")
-        .arg(format!(
-            "curl -sSLo - {} | tar xOzf - {}",
-            ICON_URL, ICON_PATH
-        ))
-        .output()
-        .map(|x| String::from_utf8(x.stdout).expect("source file is not UTF-8"))
-        .expect("could not download icons");
+    let icon_svg_paths = fs::read_to_string("iconSvgPaths.js").expect("Cannot read file");
```
Comment (cecton) [inline yewprint/build.rs:12]: I personally use only lower case in expect messages. Can you keep the consistency?

File: .github/workflows/pr.yml [hunk 5]
```diff
@@ -83,3 +83,22 @@ jobs:
...
+    runs-on: windows-latest
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
+
+      - name: Install target wasm32-unknown-unknown
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          target: wasm32-unknown-unknown
```
Comment (cecton) [inline .github/workflows/pr.yml:98]: Adding the rust-toolchain file will remove the need to do this entire step, remember?

File: .github/workflows/pr.yml [hunk 6]
```diff
@@ -83,3 +83,22 @@ jobs:
           | File name | Old Size | New Size |
           | --- | --- | --- |
           | public/wasm_bg.wasm | ${{needs.build-main.outputs.size}} | ${{needs.build.outputs.size}} |
+
+  build-windows:
+    needs: [build]
```
Comment (cecton) [inline .github/workflows/pr.yml:88]: I doubt this job is a dependency of anything. It can run on its own it is not waiting for the artifacts of another job.

File: .github/workflows/pr.yml [hunk 7]
```diff
@@ -83,3 +83,22 @@ jobs:
...
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          target: wasm32-unknown-unknown
+
+      - name: cargo test (windows)
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
+          args: --workspace
```
Comment (cecton) [inline .github/workflows/pr.yml:104]: We need to build for the correct architecture.

File: .github/workflows/pr.yml (hunk 5 above)
Comment (yozhgoor) [reply .github/workflows/pr.yml:98]: Added

File: .github/workflows/pr.yml [hunk 8]
```diff
@@ -83,3 +94,4 @@ jobs:
           | File name | Old Size | New Size |
           | --- | --- | --- |
           | public/wasm_bg.wasm | ${{needs.build-main.outputs.size}} | ${{needs.build.outputs.size}} |
+
```
Comment (cecton) [inline .github/workflows/pr.yml:97]: Unneeded diff

File: .github/workflows/pr.yml [hunk 9]
```diff
@@ -36,6 +36,17 @@ jobs:
         NETLIFY_SITE_ID: ${{ secrets.NETLIFY_SITE_ID }}
       timeout-minutes: 1
 
+  build-windows:
+    runs-on: windows-latest
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
+
```
Comment (cecton) [inline .github/workflows/pr.yml:44]: The other jobs don't have a space between the steps

File: .github/workflows/pr.yml [hunk 10]
```diff
@@ -36,6 +36,17 @@ jobs:
...
+      - name: cargo test (windows)
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
```
Comment (cecton) [inline .github/workflows/pr.yml:48]: Still need to target the correct architecture

File: yewprint/build.rs [hunk 11]
```diff
@@ -4,21 +4,17 @@ use std::collections::HashSet;
...
 use std::fs;
 use std::path::Path;
-use std::process::Command;
 
-const ICON_URL: &str = "https://registry.npmjs.org/@blueprintjs/icons/-/icons-3.19.0.tgz";
-const ICON_PATH: &str = "package/lib/esnext/generated/iconSvgPaths.js";
+/*
+If the icon file (iconSvgPaths.js) need an update, you can download
+an npm archive at this url:
+"https://registry.npmjs.org/@blueprintjs/icons/-/icons-3.19.0.tgz";
```
Comment (cecton) [inline yewprint/build.rs:11]: looks like you copy-pasted with the " and ; lol

File: .github/workflows/pr.yml (hunk 10 above)
Comment (cecton) [reply .github/workflows/pr.yml:47]: I don't think it is possible with `test`. You will need to just build (as I asked initially)

Comment (cecton) [reply .github/workflows/pr.yml:47]: (I wasn't very clear initially lol)

File: yewprint/build.rs [hunk 12]
```diff
@@ -4,21 +4,17 @@ use std::collections::HashSet;
...
 
-const ICON_URL: &str = "https://registry.npmjs.org/@blueprintjs/icons/-/icons-3.19.0.tgz";
-const ICON_PATH: &str = "package/lib/esnext/generated/iconSvgPaths.js";
+/*
+If the icon file (iconSvgPaths.js) need an update, you can download
+an npm archive at this url:
+https://registry.npmjs.org/@blueprintjs/icons/-/icons-3.19.0.tgz
+after that you can extract the file following this path:
+package/lib/esnext/generated/iconSvgPaths.js
+*/
```
Comment (cecton) [inline yewprint/build.rs:14]: Ok now I think you can convert this to module doc AND explain what this file does :grin:

File: .github/workflows/pr.yml [hunk 13]
```diff
@@ -36,6 +36,22 @@ jobs:
...
+  build-windows:
+    runs-on: windows-latest
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
+      - name: cargo build (windows)
+        uses: actions-rs/cargo@v1
+        with:
+          command: build
+          args: --target x86_64-pc-windows-pc-msvc
```
Comment (cecton) [inline .github/workflows/pr.yml:48]: Still not the correct architecture :grin:

File: .github/workflows/pr.yml [hunk 14]
```diff
@@ -36,6 +36,22 @@ jobs:
...
+      - name: cargo test (windows)
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
```
Comment (cecton) [inline .github/workflows/pr.yml:52]: No need to run the test anymore (there are none anyway).

Even if we had tests. They would already run on the Linux build. And in this type of project, tests need to run in a browser most of the time so... Just remove for now.

File: yewprint/build.rs (hunk 12 above)
Comment (yozhgoor) [reply yewprint/build.rs:14]: owh damn, lets try

File: .github/workflows/pr.yml [hunk 15]
```diff
@@ -36,6 +36,17 @@ jobs:
...
+  build-windows:
+    runs-on: windows-latest
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v2
+      - name: cargo build (windows)
+        uses: actions-rs/cargo@v1
+        with:
+          command: build
+          args: --target x86_64-pc-windows-msvc
```
Comment (cecton) [inline .github/workflows/pr.yml:48]: yewprint is a library in WASM for the browser... so the target is wasm, see? XD

File: yewprint/build.rs [hunk 16]
```diff
@@ -1,24 +1,21 @@
+//! This build script parses the svg paths located in iconSvgPaths.js
+//! to generate the IconName enum and all the icons.
+//!
+//! If the js file need an update, you can download a npm archive at this url:
+//! https://registry.npmjs.org/@blueprintjs/icons/-/icons-3.19.0.tgz
+//!
+//! after that you can extract the file following this path:
```
Comment (cecton) [inline yewprint/build.rs:7]:
```suggestion
//! After that you can extract the file following this path:
```
